Config file implementation and added `snakefmt` and `dry run` tests documentation (#5)
WayScience/CytoPipe: README.md

# Table of contents

- [Table of contents](#table-of-contents)
- [CytoPipe](#cytopipe)
- [Formatting](#formatting)
  - [smkfmt](#smkfmt)
- [Testing](#testing)
  - [Dry Runs](#dry-runs)

> # CytoPipe

Reviewer: This should be line 1!

We should also add a quick description of the package.

Maybe something like:

"Reproducible pipelines for processing high-dimensional systems morphology data with snakemake 🐍 "

Reviewer: TOC can go after this short description :)

Author: fixed! : https://github.com/WayScience/CytoPipe/pull/5/commits/7564f7a4b657c4dd35ce6e0d8705d305845f0926

> # Formatting

Reviewer: I think all headings after the first `# CytoPipe` should be `H2` headings:

AKA `## Formatting`

```suggestion
## Formatting
```

Reviewer: Please fix all headings below as follows

> ## smkfmt

Reviewer: This says `smkfmt` but below you write `snakefmt` - just checking if this is correct!

Author: Typo fixed! : https://github.com/WayScience/CytoPipe/pull/5/commits/7564f7a4b657c4dd35ce6e0d8705d305845f0926

`snakefmt` is a `snakemake` code formatter that is developed and maintained by the developers of `snakemake`.

In order to use `snakefmt`, `snakemake` must be installed, which can be found here: [snakemake-installation](https://snakemake.readthedocs.io/en/stable/getting_started/installation.html)

`snakefmt` installation processes can be found in the [snakefmt](https://github.com/snakemake/snakefmt#install) repository.

> 

Reviewer: You might consider adding python formatting style guide (aka `black`)

Author: added on : https://github.com/WayScience/CytoPipe/pull/5/commits/7564f7a4b657c4dd35ce6e0d8705d305845f0926

# Testing

## Dry Runs

> `Snakemake`'s `dry-runs` is a simple test that ensures that all the required inputs and expected outputs are satisfied. No execution is conducted therefore any potential errors found within the executables **will not be captured.**

Reviewer: ```suggestion
`Snakemake`'s `dry-runs` is a simple test that ensures that all the required inputs and expected outputs are satisfied.
No execution is conducted, therefore the dry run **will not capture** any potential errors found within the executables.
```

Two suggestions baked in here:

1. Please use [one sentence per line](https://news.ycombinator.com/item?id=4642395)
2. Edited second sentence from passive to active voice :)

Below is an example on how to conduct a `dry-run` on `snakemake`:

```text
snakemake -n
```

> Using the `-n`, by default, will conduct a `dry-run` to all rules that exists within the workflow

Reviewer: ```suggestion
Using the `-n`, by default, will conduct a `dry-run` to all rules that exists within the workflow.
```

> One can specify a specific rule to conduct a `dry-run` test by inputting the name of the rule after the `-n` flag . This will only check all the required inputs and expected outputs that are specific to the specified rule.

Reviewer: ```suggestion
One can specify a specific rule to conduct a `dry-run` test by inputting the name of the rule after the `-n` flag.
This will only check all the required inputs and expected outputs that are specific to the specified rule.
```

Neat!

> The example below executes a `dry-run` test to only the `annotate` rule

Reviewer: ```suggestion
The example below executes a `dry-run` test to only the `annotate` rule.
```

```text
snakemake -n annotate
```

output:

```text
Building DAG of jobs...
Job stats:
job          count    min threads    max threads
---------  -------  -------------  -------------
aggregate        1              1              1
annotate         1              1              1
total            2              1              1

[Mon Apr 18 10:42:55 2022]
rule aggregate:
    input: data/SQ00014614.sqlite, data/barcode_platemap.csv, data/metadata
    output: results/preprocessing/SQ00014614_cell_counts.tsv, results/preprocessing/SQ00014614_aggregate.csv.gz
    jobid: 1
    resources: tmpdir=/tmp

[Mon Apr 18 10:42:55 2022]
rule annotate:
    input: data/barcode_platemap.csv, results/preprocessing/SQ00014614_aggregate.csv.gz
    output: results/preprocessing/SQ00014614_augmented.csv.gz
    jobid: 0
    resources: tmpdir=/tmp

Job stats:
job          count    min threads    max threads
---------  -------  -------------  -------------
aggregate        1              1              1
annotate         1              1              1
total            2              1              1

> 

Reviewer: Please remove the extra blank line

Author: Fix in : https://github.com/WayScience/CytoPipe/pull/5/commits/7564f7a4b657c4dd35ce6e0d8705d305845f0926

```

> The resulting output is a summary of the `dry-run` and it shows that two rules were tested despite only specifying one. This is due to the `aggregate` rule creating an output that is a required input for the `annotate` rule; therefore, indicating that the `aggregate` rule is a requirement for the `annotate` rule .

Reviewer: ```suggestion
The resulting output is a summary of the `dry-run` and it shows that two rules were tested despite only specifying one.
This is due to the `aggregate` rule creating an output that is a required input for the `annotate` rule; therefore, indicating that the `aggregate` rule is a requirement for the `annotate` rule.
```
